### ISSUE TITLE
[1.6.2-beta.2] Babel build fails weirdly (test case included)

### ISSUE THREAD

**Issue by houshuang** [meteor/meteor#9571]
Minimal repro repo: https://github.com/houshuang/meteor-crash-babel-repro

Putting a single file with this content in the Meteor directory, and trying to run Meteor fails with a big Babel error:

```js
export default class Session {
  constructor(session) {}

  @action
  setTimes = session => {
    this.startedAt = session.startedAt;
  };

  close = () => {};
}
```
The weird thing is that removing any word in this file seems to fix the error, so this seems like a very strange bug. 

This is .babelrc:

```js
{
  "plugins": [
    "@babel/plugin-proposal-decorators",
    ["@babel/plugin-proposal-class-properties", { "loose" : true }]
  ],
  "presets": [
    "@babel/preset-env"
  ]
}
```

Note that running Babel 7 with the same plugins and .babelrc on session.js works perfectly, only Meteor fails.

Error:

```js
=> Errors prevented startup:

   While processing files with ecmascript (for target web.browser):

   /Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/modification.js:84:11:
   We don't know what to do with this node type. We were previously a Statement but we can't fit
   in here?
   at NodePath.insertBefore
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/modification.js:84:11)
   at PluginPass.Class
   (/Users/stian/src/f/node_modules/@babel/plugin-proposal-class-properties/lib/index.js:225:14)
   at newFn
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/visitors.js:223:21)
   at NodePath._call
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:64:19)
   at NodePath.call
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:38:17)
   at NodePath.visit
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:99:12)
   at TraversalContext.visitQueue
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:139:18)
   at TraversalContext.visitSingle
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:98:19)
   at TraversalContext.visit
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:180:19)
   at Function.traverse.node
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/index.js:76:17)
   at NodePath.visit
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:106:18)
   at TraversalContext.visitQueue
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:139:18)
   at TraversalContext.visitMultiple
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:93:17)
   at TraversalContext.visit
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:178:19)
   at Function.traverse.node
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/index.js:76:17)
   at NodePath.visit
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/path/context.js:106:18)
   at TraversalContext.visitQueue
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:139:18)
   at TraversalContext.visitSingle
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:98:19)
   at TraversalContext.visit
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/context.js:180:19)
   at Function.traverse.node
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/index.js:76:17)
   at traverse
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/traverse/lib/index.js:46:12)
   at transformFile
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transformation/index.js:107:27)
   at runSync
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transformation/index.js:35:3)
   at transformFromAstSync
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transform-ast-sync.js:16:38)
   at Object.transformFromAst
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/@babel/core/lib/transform-ast.js:20:68)
   at transform
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:71:30)
   at compile
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:82:14)
   at Cache.Cp.get
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/cache.js:101:26)
   at Object.compile
   (/Users/stian/.meteor/packages/ecmascript/.0.10.0.1xodfw8.8ksg++os+web.browser+web.cordova/plugin.compile-ecmascript.os/npm/node_modules/meteor/babel-compiler/node_modules/meteor-babel/index.js:46:23)
   at Object.compile (packages/babel-compiler.js:51:29)
   at packages/babel-compiler.js:186:22
   at Function.time (/tools/tool-env/profile.js:305:10)
   at profile (packages/babel-compiler.js:229:20)
   at BabelCompiler.BCp.processOneFileForTarget (packages/babel-compiler.js:185:20)
   at BabelCompiler.<anonymous> (packages/babel-compiler.js:111:26)
   at Array.forEach (<anonymous>)
   at BabelCompiler.BCp.processFilesForTarget (packages/babel-compiler.js:110:14)
```

**Comment by benjamn** [comment]
A common problem we've seen with supporting custom `.babelrc` files is that people add presets and plugins to `.babelrc` that are already covered by Meteor's [`babel-preset-meteor`](https://github.com/meteor/babel-preset-meteor) (with some additional plugins coming from https://github.com/meteor/babel/blob/master/options.js).

In this case, `@babel/plugin-proposal-decorators` is necessary and worth including in `.babelrc`, because Meteor does not provide it already. That proposal has changed significantly over time, in breaking ways, so we haven't felt comfortable encouraging its use. Developers who accept the risk of it changing again in the future are free to put it in their `.babelrc` files, though!

However, `@babel/plugin-proposal-class-properties` is already provided by Meteor, and `babel-preset-env` is unnecessary because Meteor already provides an appropriate set of core plugins for each environment (Node 8, modern browsers, and legacy browsers). You can and probably should remove both of those packages from `.babelrc`, and instead add any plugins individually that are not provided already by Meteor.

With that said, after playing around with your reproduction (thanks!), `babel-preset-env` appears not to cause any problems here. The duplicate `@babel/plugin-proposal-class-properties` seems to be the source of the problem, and probably does need to be removed. For whatever reason, running that plugin twice is not only a waste of compilation time, but it also triggers this internal bug in Babel. The reason this appears to work in a non-Meteor project is likely that `@babel/plugin-proposal-class-properties` appears only once in the configuration.

While it might be nice to fix this problem somehow, I can't promise that all `.babelrc` configurations are guaranteed to work, and instead I would recommend adding specific plugins as needed, rather than using large presets (even though it seems to work). Your Babel compile times will improve, and you'll be less likely to experience weird plugin ordering problems like this in the future.

**Comment by houshuang** [comment]
Thank you so much for your detailed feedback. My problem was thinking that it was OK to use the same .babelrc plugin to compile all the packages (which are just ES6 with Babel) and the main Meteor package - I now realize these need to be different.

For me the issue is resolved - the only improvement could be some documentation of this stuff in the core docs (or perhaps it's already there). If you are planning to make some further improvements, you can keep this issue open, otherwise you can close it.

Thanks again!

**Comment by hwillson** [comment]
Thanks for the feedback @houshuang - adding an extra mention about this in the docs is a great idea. I've opened https://github.com/meteor/guide/issues/690 to track this (for the Guide). Closing here - thanks again!
